### PR TITLE
EE-730

### PR DESCRIPTION
**Pull request by jorizci**
- Database password encryption
- File Manager now pulls events from File-re-encryption job complete to know when a file has been re-encrypted
- Fixed database PK for the historic download box generation

### CODE REVIEW

File: file-manager/src/main/java/uk/ac/ebi/ega/ingestion/file/manager/services/DownloadBoxJobService.java [hunk 1]
```diff
@@ -89,8 +93,11 @@ public DownloadBoxJob createJob(DownloadBoxJob downloadBoxJob) {
     }
 
     @Override
-    public synchronized void finishFileJob(long id) throws FileJobNotFound {
+    @Transactional(transactionManager = "fileManager_transactionManager")
+    public synchronized void finishFileJob(String id, String message, LocalDateTime start, LocalDateTime finish)
+            throws FileJobNotFound {
         DownloadBoxFileJob downloadBoxFileJob = boxFileJobRepository.findById(id).orElseThrow(FileJobNotFound::new);
+
```
Comment (ashutoshebi) [inline file-manager/src/main/java/uk/ac/ebi/ega/ingestion/file/manager/services/DownloadBoxJobService.java]: As discussed, use of synchronized needs to be revisited.

File: file-manager/src/main/java/uk/ac/ebi/ega/ingestion/file/manager/services/MailingService.java [hunk 2]
```diff
@@ -59,4 +84,11 @@ public void sendSimpleMessage(String to, String subject, String text) {
         mailSender.send(message);
     }
 
+    @Override
+    public void reportError(String subject, Exception e) {
+        StringWriter stringWriter = new StringWriter();
+        e.printStackTrace(new PrintWriter(stringWriter));
```
Comment (ashutoshebi) [inline file-manager/src/main/java/uk/ac/ebi/ega/ingestion/file/manager/services/MailingService.java]: StackTrace is expected to be logged in Logger. Not sure about PrintWriter.

File: file-manager/src/main/java/uk/ac/ebi/ega/ingestion/file/manager/utils/FileExtensionUtils.java [hunk 3]
```diff
@@ -92,9 +94,18 @@ public static String getFileExtension(String filenameWithExtensions) {
...
     }
 
+
+    public static String getEncryptionExtension(String fileExtensionWithEncryption) {
+        if (fileExtensionWithEncryption.endsWith("gpg") || fileExtensionWithEncryption.endsWith("cip")) {
+            return fileExtensionWithEncryption.substring(fileExtensionWithEncryption.length() - 3);
+        }
+        return null;
+    }
+
```
Comment (ashutoshebi) [inline file-manager/src/main/java/uk/ac/ebi/ega/ingestion/file/manager/utils/FileExtensionUtils.java]: I assume, we are making sure of non null references. We are not checking for null check it is good if we are passing at least empty string or objects.

File: encryption-core/src/main/java/uk/ac/ebi/ega/encryption/core/encryption/JdkEncryptionAlgorithm.java [hunk 4]
```diff
@@ -39,6 +44,32 @@ public InputStream decrypt(InputStream inputStream, char[] password) throws Algo
...
 
+    public byte[] encrypt(char[] password, byte[] content) throws AlgorithmInitializationException {
+        try (final ByteArrayOutputStream outputStream = new ByteArrayOutputStream(content.length)) {
+            initializeWrite(password, outputStream);
+            try (OutputStream encryptStream = new CipherOutputStream(outputStream, getCipher(Cipher.ENCRYPT_MODE))) {
+                encryptStream.write(content);
+                encryptStream.flush();
+            }
+            outputStream.flush();
+            outputStream.close();
```
[the remark this answers is not in the document]
Comment (jorizci) [reply encryption-core/src/main/java/uk/ac/ebi/ega/encryption/core/encryption/JdkEncryptionAlgorithm.java]: This was done intentionately, as we want to be sure that wrapped output stream have been flushed and closed properly BEFORE retrieving the byte array from it. There are some instances that more information is written only when close is invoked.

File: encryption-core/src/main/java/uk/ac/ebi/ega/encryption/core/services/PasswordEncryptionService.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import uk.ac.ebi.ega.encryption.core.utils.io.IOUtils;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+public class PasswordEncryptionService implements IPasswordEncryptionService {
+
+    private char[] passwordKey;
+
+    private AesCbcOpenSSL aesCbcOpenSSL;
+
```
[the remark this answers is not in the document]
Comment (jorizci) [reply encryption-core/src/main/java/uk/ac/ebi/ega/encryption/core/services/PasswordEncryptionService.java]: This is intentional, usually I would concur (in fact I used to follow that practise), but on classes that can be autowired force the final can block future ways of autowiring things.

File: encryption-core/src/main/java/uk/ac/ebi/ega/encryption/core/utils/io/IOUtils.java [hunk 6]
```diff
@@ -56,4 +59,19 @@ public static void skipFully(InputStream inputStream, long totalBytes) throws IO
...
+    }
+
+    private static void copy(InputStreamReader reader, CharArrayWriter writer) throws IOException {
+        char[] buffer = new char[1024];
+        int totalRead;
+        while ((totalRead = reader.read(buffer)) != -1) {
+            writer.write(buffer, 0, totalRead);
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (jorizci) [reply encryption-core/src/main/java/uk/ac/ebi/ega/encryption/core/utils/io/IOUtils.java]: It is responsability of whoever uses it, this is only intended as a helper and depending on the use case you want it to be closed or not.

File: file-manager/src/main/java/uk/ac/ebi/ega/ingestion/file/manager/services/MailingService.java (hunk 2 above)
Comment (jorizci) [reply file-manager/src/main/java/uk/ac/ebi/ega/ingestion/file/manager/services/MailingService.java]: That converts the stacktrace into a string that can be send through mail.

File: file-manager/src/main/java/uk/ac/ebi/ega/ingestion/file/manager/utils/FileExtensionUtils.java (hunk 3 above)
Comment (jorizci) [reply file-manager/src/main/java/uk/ac/ebi/ega/ingestion/file/manager/utils/FileExtensionUtils.java]: fileExtensionWithEncryption should alwats be not null (unless somebody is screwing with the system data in which case it should explode)

**Pull request merged**